### PR TITLE
Implement exact source (xs) for magnet URIs

### PR DESCRIPTION
**Pull request by Sebmaster**
Since I saw #714 stalled a bit, I wanted to check in if no one's working on this and started off with a minimal working prototype.

I know I still need to move the fixture over (also don't rely on webtorrent.io, but create a local server instead) to the other repo, but does this look okay / is anyone else working on this?

### CODE REVIEW

File: lib/torrent.js [hunk 1]
```diff
@@ -269,6 +271,34 @@ Torrent.prototype._onParsedTorrent = function (parsedTorrent) {
...
+      url: parsedTorrent.xs,
+      method: 'GET',
+      headers: {
+        'user-agent': 'WebTorrent (http://webtorrent.io)'
+      }
+    }
+    try {
+      this._xsRequest = get.concat(opts, function (err, res, torrent) {
+        if (err || !torrent.length) return
+
```
Comment (feross) [inline lib/torrent.js]: Should check that `res.statusCode === 200`

**Comment by feross** [comment]
Just took a quick look. Looks pretty good to me. I don't think we to/can easily support ftp:// so I would remove that test case.

**Comment by Sebmaster** [comment]
Fixed. 

> I don't think we to/can easily support ftp:// so I would remove that test case.

I added that mainly to check if unsupported protocols would crash the process, since the spec says to silently ignore.

**Comment by feross** [comment]
I think we should check for http:// https:// explicitly before passing into `simple-get`. That would be more reliable and we could get rid of the try-catch around `simple-get`, right?

**Comment by Sebmaster** [comment]
Done.

> That would be more reliable and we could get rid of the try-catch around `simple-get`, right?

For the protocol case yeah, just hope there's nothing else `http` will throw on...

**Comment by Sebmaster** [comment]
Okay, also migrated the tests over to a local server and reused the fixtures from `webtorrent-fixtures`. Should be ready (for xs, at least).

**Comment by feross** [comment]
Thanks -- will take a look after getting WebTorrent Desktop v0.5 out.

**Comment by Sebmaster** [comment]
I just checked the malformed URL case some more - a URL like `http://example.org:2455665555/` would now crash, which isn't great, so I'm thinking leaving the try / catch there would be better (either instead, or in addition to the protocol check).

File: lib/torrent.js [hunk 2]
```diff
@@ -269,6 +271,38 @@ Torrent.prototype._onParsedTorrent = function (parsedTorrent) {
       self._onListening()
     })
   }
+
+  if (parsedTorrent.xs && !self.info) {
+    // TODO: Could try multiple in parallel here
+    var src = parsedTorrent.xs instanceof Array ? parsedTorrent.xs[0] : parsedTorrent.xs
```
Comment (feross) [inline lib/torrent.js]: `Array.isArray()` is the right way to check if something is an Array.

File: lib/torrent.js [hunk 3]
```diff
@@ -269,6 +271,38 @@ Torrent.prototype._onParsedTorrent = function (parsedTorrent) {
...
+    if (isHTTP) {
+      var opts = {
+        url: src,
+        method: 'GET',
+        headers: {
+          'user-agent': 'WebTorrent (http://webtorrent.io)'
+        }
+      }
+      this._xsRequest = get.concat(opts, function (err, res, torrent) {
+        if (err || res.statusCode !== 200 || !torrent.length) return
```
Comment (feross) [inline lib/torrent.js]: No need to check for `!torrent.length` since that will make `parseTorrent` throw below anyway.

Comment (feross) [inline lib/torrent.js]: Would be nice to add a log here with `debug` for debugging purposes.

File: lib/torrent.js [hunk 4]
```diff
@@ -269,6 +271,38 @@ Torrent.prototype._onParsedTorrent = function (parsedTorrent) {
...
+
+        var parsedTorrent
+        try {
+          parsedTorrent = parseTorrent(torrent)
+        } catch (err) {}
+
+        if (parsedTorrent) {
+          if (parsedTorrent.infoHash !== self.infoHash) {
+            return
```
Comment (feross) [inline lib/torrent.js]: Debug log here too

File: lib/torrent.js [hunk 5]
```diff
@@ -371,6 +405,10 @@ Torrent.prototype._onListening = function () {
  */
 Torrent.prototype._onMetadata = function (metadata) {
   var self = this
+  if (self._xsRequest) {
+    self._xsRequest.abort()
+    self._xsRequest = null
+  }
```
Comment (feross) [inline lib/torrent.js]: This should go after the next line.

File: test/node/download-metadata.js [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+}
+
+test('Download metadata for magnet URI with xs parameter', function (t) {
+  t.plan(2)
+
+  var client = new WebTorrent({ dht: false, tracker: false })
+
+  createServer(fixtures.leaves.torrent, function (url, next) {
+    client.add(fixtures.leaves.magnetURI + '&xs=' + encodeURIComponent(url), function (torrent) {
+      t.equal(torrent.name, 'Leaves of Grass by Walt Whitman.epub')
```
Comment (feross) [inline test/node/download-metadata.js]: The name is also in the magnet URI as the `&dn` parameter. Might be better to check for another parameter.

File: test/node/download-metadata.js [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+  })
+})
+
+test('Download metadata magnet URI with unsupported protocol in xs parameter', function (t) {
+  t.plan(2)
+
+  var client = new WebTorrent({ dht: false, tracker: false })
+  client.add(fixtures.leaves.magnetURI + '&xs=' + encodeURIComponent('invalidurl:example'))
+  setTimeout(function () {
+    t.ok(true, 'no crash')
```
Comment (feross) [inline test/node/download-metadata.js]: No need for this. The tests will fail if there's a thrown exception. We should add listeners for 'error' and 'warning' events to all these tests though.

**Comment by Sebmaster** [comment]
I totally fixed all of those just now :smile: 

Also what about the try-catch situation with malformed urls?

**Comment by feross** [comment]
Ah, thanks. I'm fixing all that up too. Sorry, should have communicated that better.

**Comment by feross** [comment]
I'm adding support for multiple &xs= params too.

File: test/node/download-metadata.js (hunk 7 above)
Comment (feross) [inline test/node/download-metadata.js]: Nevermind, I see what you were doing.

**Comment by feross** [comment]
@Sebmaster See https://github.com/feross/webtorrent/pull/809

**Pull request merged**